src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpChangeNamespaceService.cs

```diff
@@ -135,89 +192,163 @@ protected override SyntaxList<MemberDeclarationSyntax> GetMemberDeclarationsInCo
+        ///   inside (i.e. all members are declared in global namespace)
+        /// </summary>
+        protected override async Task<SyntaxNode> TryGetApplicableContainerFromSpanAsync(Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var compilationUnit = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false) as CompilationUnitSyntax;
+            SyntaxNode container = null;
 
-            // Move everything to global namespace
-            if (IsGlobalNamespace(targetNamespaceParts))
+            // Empty span means that user wants to move all types declared in the document to a new namespace.
+            // This action is only supported when everything in the document is declared in global namespace,
+            // which we use the number of namespace declaration nodes to decide.
+            if (span.IsEmpty)
             {
-                var (namespaceOpeningTrivia, namespaceClosingTrivia) = 
-                    GetOpeningAndClosingTriviaOfNamespaceDeclaration(namespaceDeclaration);
-                var members = namespaceDeclaration.Members;
-                var eofToken = compilationUnit.EndOfFileToken
-                    .WithAdditionalAnnotations(WarningAnnotation);
-
-                // Try to preserve trivia from original namesapce declaration.
-                // If there's any member inside the declaration, we attach them to the 
-                // first and last member, otherwise, simply attach all to the EOF token.
-                if (members.Count > 0)
+                var hasNamespaceDecl = compilationUnit.DescendantNodes(IsCompilationUnitOrNamespaceDeclaration)
+                    .OfType<NamespaceDeclarationSyntax>().Any();
+
+                if (hasNamespaceDecl)
                 {
-                    var first = members.First();
-                    var firstWithTrivia = first.WithPrependedLeadingTrivia(namespaceOpeningTrivia);
-                    members = members.Replace(first, firstWithTrivia);
+                    return null;
+                }
+
+                container = compilationUnit;
+            }
+            else
+            {
+                // Otherwise, the span should contain a namespace declaration node, which must be the only one
+                // in the entire syntax spine to enable the change namespace operation.
+                var node = compilationUnit.FindNode(span, getInnermostNodeForTie: true);
 
-                    var last = members.Last();
-                    var lastWithTrivia = last.WithAppendedTrailingTrivia(namespaceClosingTrivia);
-                    members = members.Replace(last, lastWithTrivia);
+                var namespaceDecl = node.AncestorsAndSelf().OfType<NamespaceDeclarationSyntax>().SingleOrDefault();
+                if (namespaceDecl?.Name.GetDiagnostics().All(diag => diag.DefaultSeverity != DiagnosticSeverity.Error) != true)
```

Comment:
the `?.` and `All` and `!= true` are jsut super confusing.  First, if you don't have a namespace, handle htat.  Then just check the diagnostics of the name.

Reply:
Fixed